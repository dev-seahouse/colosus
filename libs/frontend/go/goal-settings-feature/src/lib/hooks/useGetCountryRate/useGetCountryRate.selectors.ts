import { useQuery } from '@tanstack/react-query';

import { getCountryRateQuery } from './useGetCountryRate';
import type { GetCountryRateData } from './useGetCountryRate';

/**
 * hook to get life expectancy and interest rate data
 */
export const useSelectLifeExpectancyAndInterestQuery = () => {
  return useQuery({
    ...getCountryRateQuery('US'),
    select: (data: GetCountryRateData) => {
      if (!data || !data.length) return null;

      return {
        lifeExpectancyMale: data[0].lifeExpectancyMale,
        lifeExpectancyFemale: data[0].lifeExpectancyFemale,
        annualisedSavingsAcctIntR: data[0].savingsAccountInterestRate,
        annualisedInflationRate: data[0].inflationRateLongTerm,
      };
    },
  });
};

/**
 * hook to get life expectancy and interest rate data
 */
export const useSelectInflationRateQuery = () => {
  return useQuery({
    ...getCountryRateQuery('US'),
    select: (data: GetCountryRateData) => {
      if (!data || !data.length) return 0;

      return data[0].inflationRateLongTerm;
    },
  });
};
