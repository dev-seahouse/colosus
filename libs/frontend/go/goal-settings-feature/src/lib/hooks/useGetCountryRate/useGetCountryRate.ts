import type { QueryClient } from '@tanstack/react-query';
import { useQuery } from '@tanstack/react-query';
import { BambuApiLibraryIntegrationCountryMetadataApi } from '@bambu/api-client';
import type {
  GetCountryRatesResponseDto,
  GetCountryRatesRequestDto,
} from '@bambu/api-client';

export type GetCountryRateData = GetCountryRatesResponseDto;

export const getCountryRateQuery = (
  countryCode: GetCountryRatesRequestDto['countryAlpha2Code']
) => ({
  queryKey: ['getCountryRate', countryCode],
  queryFn: async () => {
    const bambuApiLibraryIntegrationCountryMetadataApi =
      new BambuApiLibraryIntegrationCountryMetadataApi();
    const res =
      await bambuApiLibraryIntegrationCountryMetadataApi.getCountryRates({
        countryAlpha2Code: countryCode,
      });

    return res.data;
  },
});

export const getCountryRateLoader =
  (queryClient: QueryClient) => async (): Promise<GetCountryRateData> => {
    const query = getCountryRateQuery('US');

    // ⬇️ return data or fetch it
    return (
      queryClient.getQueryData(query.queryKey) ??
      (await queryClient.fetchQuery(query))
    );
  };

export interface UseGetCountryRateOptions {
  initialData?: Partial<GetCountryRateData>;
}

/**
 * hook to get country rate data
 */
export const useGetCountryRate = ({
  initialData,
}: UseGetCountryRateOptions = {}) => {
  return useQuery({
    ...getCountryRateQuery('US'),
    initialData,
  });
};

export default useGetCountryRate;
