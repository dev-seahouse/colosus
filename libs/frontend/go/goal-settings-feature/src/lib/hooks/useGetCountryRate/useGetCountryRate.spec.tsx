import {
  renderHook,
  queryClientWrapper,
  waitFor,
} from '@bambu/react-test-utils';

import useGetCountryRate from './useGetCountryRate';

describe('useGetCountryRate', () => {
  it('should render successfully', async () => {
    const { result } = renderHook(() => useGetCountryRate(), {
      wrapper: queryClientWrapper,
    });

    await waitFor(() => expect(result.current.isSuccess).toBe(true));

    expect(result.current.data).toMatchSnapshot();
  });
});
