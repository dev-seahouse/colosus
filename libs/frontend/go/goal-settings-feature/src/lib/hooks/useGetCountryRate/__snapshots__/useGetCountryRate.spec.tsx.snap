// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`useGetCountryRate > should render successfully 1`] = `
[
  {
    "annualWageGrowthLongTerm": 0.0024,
    "annualWageGrowthShortTerm": 0.0024,
    "averageSavingsRate": 0.081,
    "countryCode": "US",
    "dateCreated": "2019-02-17T21:55:33.000Z",
    "id": 142,
    "inflationRateLongTerm": 0.017,
    "inflationRateShortTerm": 0.0132,
    "lifeExpectancyFemale": 81,
    "lifeExpectancyMale": 76,
    "retirementAgeFemale": 66,
    "retirementAgeMale": 66,
    "savingsAccountInterestRate": 0.02,
  },
]
`;
